Build failure in VMs: Error: no such instruction: `tpause %ecx'

We are pulling the latest on master branch (but have tried other recent tagged versions) as a submodule to another project.  The work is being done in an Ubuntu 18.04 LTS VM in VMWare Workstation on a Windows x86 machine and in VMWare Fusion on an x86 macBook.  We are using gcc/g++ 11.1.0.

We get the following error message on both systems (there is no such error when not using a VM):

```
[100%] Linking CXX shared library ../../../../gnu_11.1_cxx20_64_release/libtbb.so
cd /home/smcmillan/github/gbtl/build_lums_cmake/external/oneTBB/src/tbb && /usr/bin/cmake -E cmake_link_script CMakeFiles/tbb.dir/link.txt --verbose=true
/usr/bin/c++ -fPIC -O3 -DNDEBUG -flto -fno-fat-lto-objects -Wl,--version-script=/home/smcmillan/github/gbtl/external/oneTBB/src/tbb/def/lin64-tbb.def -shared -Wl,-soname,libtbb.so.12 -o ../../../../gnu_11.1_cxx20_64_release/libtbb.so.12.7 CMakeFiles/tbb.dir/address_waiter.cpp.o CMakeFiles/tbb.dir/allocator.cpp.o CMakeFiles/tbb.dir/arena.cpp.o CMakeFiles/tbb.dir/arena_slot.cpp.o CMakeFiles/tbb.dir/concurrent_bounded_queue.cpp.o CMakeFiles/tbb.dir/dynamic_link.cpp.o CMakeFiles/tbb.dir/exception.cpp.o CMakeFiles/tbb.dir/governor.cpp.o CMakeFiles/tbb.dir/global_control.cpp.o CMakeFiles/tbb.dir/itt_notify.cpp.o CMakeFiles/tbb.dir/main.cpp.o CMakeFiles/tbb.dir/market.cpp.o CMakeFiles/tbb.dir/misc.cpp.o CMakeFiles/tbb.dir/misc_ex.cpp.o CMakeFiles/tbb.dir/observer_proxy.cpp.o CMakeFiles/tbb.dir/parallel_pipeline.cpp.o CMakeFiles/tbb.dir/private_server.cpp.o CMakeFiles/tbb.dir/profiling.cpp.o CMakeFiles/tbb.dir/rml_tbb.cpp.o CMakeFiles/tbb.dir/rtm_mutex.cpp.o CMakeFiles/tbb.dir/rtm_rw_mutex.cpp.o CMakeFiles/tbb.dir/semaphore.cpp.o CMakeFiles/tbb.dir/small_object_pool.cpp.o CMakeFiles/tbb.dir/task.cpp.o CMakeFiles/tbb.dir/task_dispatcher.cpp.o CMakeFiles/tbb.dir/task_group_context.cpp.o CMakeFiles/tbb.dir/version.cpp.o CMakeFiles/tbb.dir/queuing_rw_mutex.cpp.o  -ldl -pthread
/tmp/cclIwmtb.s: Assembler messages:
/tmp/cclIwmtb.s:16988: Error: no such instruction: `tpause %ecx'
lto-wrapper: fatal error: /usr/bin/c++ returned 1 exit status
compilation terminated.
/usr/bin/ld: error: lto-wrapper failed
collect2: error: ld returned 1 exit status
external/oneTBB/src/tbb/CMakeFiles/tbb.dir/build.make:529: recipe for target 'gnu_11.1_cxx20_64_release/libtbb.so.12.7' failed
make[3]: *** [gnu_11.1_cxx20_64_release/libtbb.so.12.7] Error 1
```

To workaround this problem, we editted include/oneapi/tbb/detail/_config.h in TBB and changed lines around 332 to look like this 

```
#if 0
#define __TBB_WAITPKG_INTRINSICS_PRESENT ((__INTEL_COMPILER >= 1900 || __TBB_GCC_VERSION >= 110000 || __TBB_CLANG_VERSION >= 120000) \
                                         && (_WIN32 || _WIN64 || __unix__ || __APPLE__) && (__TBB_x86_32 || __TBB_x86_64) && !__ANDROID__)
#else
#define __TBB_WAITPKG_INTRINSICS_PRESENT 0
#endif
```

Basically, to set __TBB_WAITPKG_INTRINSICS_PRESENT to be false.

I am sure there is a better way, but what is it?

I tried this and got the following...tail of compilation:

```
[ 90%] Building CXX object external/oneTBB/src/tbb/CMakeFiles/tbb.dir/queuing_rw_mutex.cpp.o
cd /home/smcmillan/github/gbtl/build_lums_cmake/external/oneTBB/src/tbb && /usr/bin/c++ -D__TBB_BUILD -D__TBB_USE_ITT_NOTIFY -I/home/smcmillan/github/gbtl/external/oneTBB/src/tbb/../../include -DTBB_ENABLE_IPO=OFF -O3 -DNDEBUG -flto=auto -fno-fat-lto-objects -fPIC -flifetime-dse=1 -Wall -Wextra -Werror -Wfatal-errors -mrtm -mwaitpkg -pthread -std=c++20 -MD -MT external/oneTBB/src/tbb/CMakeFiles/tbb.dir/queuing_rw_mutex.cpp.o -MF CMakeFiles/tbb.dir/queuing_rw_mutex.cpp.o.d -o CMakeFiles/tbb.dir/queuing_rw_mutex.cpp.o -c /home/smcmillan/github/gbtl/external/oneTBB/src/tbb/queuing_rw_mutex.cpp
[ 90%] Linking CXX shared library ../../../../gnu_11.1_cxx20_64_release/libtbb.so
cd /home/smcmillan/github/gbtl/build_lums_cmake/external/oneTBB/src/tbb && /usr/bin/cmake -E cmake_link_script CMakeFiles/tbb.dir/link.txt --verbose=1
/usr/bin/c++ -fPIC  -DTBB_ENABLE_IPO=OFF -O3 -DNDEBUG -flto=auto -fno-fat-lto-objects -Wl,--version-script=/home/smcmillan/github/gbtl/external/oneTBB/src/tbb/def/lin64-tbb.def -shared -Wl,-soname,libtbb.so.12 -o ../../../../gnu_11.1_cxx20_64_release/libtbb.so.12.7 CMakeFiles/tbb.dir/address_waiter.cpp.o CMakeFiles/tbb.dir/allocator.cpp.o CMakeFiles/tbb.dir/arena.cpp.o CMakeFiles/tbb.dir/arena_slot.cpp.o CMakeFiles/tbb.dir/concurrent_bounded_queue.cpp.o CMakeFiles/tbb.dir/dynamic_link.cpp.o CMakeFiles/tbb.dir/exception.cpp.o CMakeFiles/tbb.dir/governor.cpp.o CMakeFiles/tbb.dir/global_control.cpp.o CMakeFiles/tbb.dir/itt_notify.cpp.o CMakeFiles/tbb.dir/main.cpp.o CMakeFiles/tbb.dir/market.cpp.o CMakeFiles/tbb.dir/misc.cpp.o CMakeFiles/tbb.dir/misc_ex.cpp.o CMakeFiles/tbb.dir/observer_proxy.cpp.o CMakeFiles/tbb.dir/parallel_pipeline.cpp.o CMakeFiles/tbb.dir/private_server.cpp.o CMakeFiles/tbb.dir/profiling.cpp.o CMakeFiles/tbb.dir/rml_tbb.cpp.o CMakeFiles/tbb.dir/rtm_mutex.cpp.o CMakeFiles/tbb.dir/rtm_rw_mutex.cpp.o CMakeFiles/tbb.dir/semaphore.cpp.o CMakeFiles/tbb.dir/small_object_pool.cpp.o CMakeFiles/tbb.dir/task.cpp.o CMakeFiles/tbb.dir/task_dispatcher.cpp.o CMakeFiles/tbb.dir/task_group_context.cpp.o CMakeFiles/tbb.dir/version.cpp.o CMakeFiles/tbb.dir/queuing_rw_mutex.cpp.o  -ldl -pthread 
/tmp/ccXIbG4k.s: Assembler messages:
/tmp/ccXIbG4k.s:9166: Error: no such instruction: `tpause %ecx'
/tmp/ccXIbG4k.s:10162: Error: no such instruction: `tpause %ecx'
make[4]: *** [/tmp/cc9ICkYG.ltrans2.ltrans.o] Error 1
make[4]: *** Waiting for unfinished jobs....
/tmp/ccUJHRYl.s: Assembler messages:
/tmp/ccUJHRYl.s:16988: Error: no such instruction: `tpause %ecx'
make[4]: *** [/tmp/cc9ICkYG.ltrans0.ltrans.o] Error 1
lto-wrapper: fatal error: make returned 2 exit status
compilation terminated.
/usr/bin/ld: error: lto-wrapper failed
collect2: error: ld returned 1 exit status
external/oneTBB/src/tbb/CMakeFiles/tbb.dir/build.make:529: recipe for target 'gnu_11.1_cxx20_64_release/libtbb.so.12.7' failed
make[3]: *** [gnu_11.1_cxx20_64_release/libtbb.so.12.7] Error 1
```

--- a/cmake/compilers/GNU.cmake
+++ b/cmake/compilers/GNU.cmake
@@ -36,17 +36,16 @@ if (NOT CMAKE_GENERATOR MATCHES "Ninja" AND NOT CMAKE_CXX_DEPENDS_USE_COMPILER)
 endif()
 
 
-# Binutils < 2.31.1 does not support the "tpause" instruction.
-# When compiling with a modern version of GCC (which does support it), but then
-# relying on an outdated assembler will fail in the assembler.
-# Here we will invoke the GNU assembler to get the version number and
+# Binutils < 2.31.1 do not support the tpause instruction.
+# When compiling with a modern version of GCC (supporting it) but
+# relying on an outdated assembler, the assembler fails.
+# The following code invokes the GNU assembler to get the version number and
 # convert it to an integer that can be used in the C++ code to compare
 # against, and disable the __TBB_WAITPKG_INTRINSICS_PRESENT macro.
-# Unfortunately, binutils only reports the MAJOR.MINOR version, without patch,
-# so the check we effectively implement is version >=2.32 (instead of >=2.31.1).
+# Binutils only report the version in the MAJOR.MINOR format, 
+# therefore the version checked is >=2.32 (instead of >=2.31.1).
 # Capturing the output in CMake can be done like below. The version information
-# will written to either stdout or stderr; for me it's stderr, but let's just
-# grab both to be sure.
+# is written to either stdout or stderr. Here, both are used. 
 execute_process(
     COMMAND ${CMAKE_CXX_COMPILER} -xc -c /dev/null -Wa,-v -o/dev/null
     OUTPUT_VARIABLE ASSEMBLER_VERSION_LINE_OUT

--- a/cmake/compilers/GNU.cmake
+++ b/cmake/compilers/GNU.cmake
@@ -55,19 +55,16 @@ execute_process(
     ERROR_STRIP_TRAILING_WHITESPACE
 )
 set(ASSEMBLER_VERSION_LINE ${ASSEMBLER_VERSION_LINE_OUT}${ASSEMBLER_VERSION_LINE_ERR})
-string(REGEX REPLACE "GNU assembler version ([0-9]+\\.[0-9]+).*" "\\1" GNU_ASSEMBLER_VERSION "${ASSEMBLER_VERSION_LINE}")
-string(FIND ${GNU_ASSEMBLER_VERSION} "." DOT_POSITION)
-if(DOT_POSITION EQUAL -1)
-    message(FATAL_ERROR "Invalid version string format: ${GNU_ASSEMBLER_VERSION}")
-endif()
+string(REGEX REPLACE "GNU assembler version ([0-9]+)\\.([0-9]+).*" "\\1" _tbb_gnu_asm_major_version "${ASSEMBLER_VERSION_LINE}")
+string(REGEX REPLACE "GNU assembler version ([0-9]+)\\.([0-9]+).*" "\\2" _tbb_gnu_asm_minor_version "${ASSEMBLER_VERSION_LINE}")
+unset(ASSEMBLER_VERSION_LINE_OUT)
+unset(ASSEMBLER_VERSION_LINE_ERR)
+unset(ASSEMBLER_VERSION_LINE)
+message(TRACE "Extracted GNU asssembler version: major=${_tbb_gnu_asm_major_version} minor=${_tbb_gnu_asm_minor_version}")
 
-# Extract the major and minor version numbers (without relying on REGEX to support older CMake versions <3.9)
-string(SUBSTRING ${GNU_ASSEMBLER_VERSION} 0 ${DOT_POSITION} GNU_ASSEMBLER_MAJOR_VERSION)
-string(SUBSTRING ${GNU_ASSEMBLER_VERSION} ${DOT_POSITION} -1 DOT_AND_MINOR_VERSION)
-string(SUBSTRING ${DOT_AND_MINOR_VERSION} 1 -1 GNU_ASSEMBLER_MINOR_VERSION)
-math(EXPR GNU_ASSEMBLER_VERSION_NUMBER  "${GNU_ASSEMBLER_MAJOR_VERSION} * 1000 + ${GNU_ASSEMBLER_MINOR_VERSION}")
-set(TBB_COMMON_COMPILE_FLAGS ${TBB_COMMON_COMPILE_FLAGS} "-DGNU_AS_VERSION=${GNU_ASSEMBLER_VERSION_NUMBER}")
-message(STATUS "GNU Assembler version: ${GNU_ASSEMBLER_VERSION}  (${GNU_ASSEMBLER_VERSION_NUMBER})")
+math(EXPR _tbb_gnu_asm_version_number  "${_tbb_gnu_asm_major_version} * 1000 + ${_tbb_gnu_asm_minor_version}")
+set(TBB_COMMON_COMPILE_FLAGS ${TBB_COMMON_COMPILE_FLAGS} "-D__TBB_GNU_ASM_VERSION=${_tbb_gnu_asm_version_number}")
+message(STATUS "GNU Assembler version: ${_tbb_gnu_asm_major_version}.${_tbb_gnu_asm_minor_version}  (${_tbb_gnu_asm_version_number})")
 
 # Enable Intel(R) Transactional Synchronization Extensions (-mrtm) and WAITPKG instructions support (-mwaitpkg) on relevant processors
 if (CMAKE_SYSTEM_PROCESSOR MATCHES "(AMD64|amd64|i.86|x86)")

--- a/cmake/compilers/GNU.cmake
+++ b/cmake/compilers/GNU.cmake
@@ -35,6 +35,40 @@ if (NOT CMAKE_GENERATOR MATCHES "Ninja" AND NOT CMAKE_CXX_DEPENDS_USE_COMPILER)
     set(TBB_MMD_FLAG -MMD)
 endif()
 
+
+# Binutils < 2.31.1 does not support the "tpause" instruction.
+# When compiling with a modern version of GCC (which does support it), but then
+# relying on an outdated assembler will fail in the assembler.
+# Here we will invoke the GNU assembler to get the version number and
+# convert it to an integer that can be used in the C++ code to compare
+# against, and disable the __TBB_WAITPKG_INTRINSICS_PRESENT macro.
+# Unfortunately, binutils only reports the MAJOR.MINOR version, without patch,
+# so the check we effectively implement is version >2.32 (instead of >=2.31.1).
+# Capturing the output in CMake can be done like below. The version information
+# will written to either stdout or stderr; for me it's stderr, but let's just
+# grab both to be sure.
+execute_process(
+    COMMAND ${CMAKE_CXX_COMPILER} -xc -c /dev/null -Wa,-v -o/dev/null
+    OUTPUT_VARIABLE ASSEMBLER_VERSION_LINE_OUT
+    ERROR_VARIABLE ASSEMBLER_VERSION_LINE_ERR
+    OUTPUT_STRIP_TRAILING_WHITESPACE
+    ERROR_STRIP_TRAILING_WHITESPACE
+)
+set(ASSEMBLER_VERSION_LINE ${ASSEMBLER_VERSION_LINE_OUT}${ASSEMBLER_VERSION_LINE_ERR})
+string(REGEX REPLACE "GNU assembler version ([0-9]+\\.[0-9]+).*" "\\1" GNU_ASSEMBLER_VERSION "${ASSEMBLER_VERSION_LINE}")
+string(FIND ${GNU_ASSEMBLER_VERSION} "." DOT_POSITION)
+if(DOT_POSITION EQUAL -1)
+    message(FATAL_ERROR "Invalid version string format: ${GNU_ASSEMBLER_VERSION}")
+endif()
+
+# Extract the major and minor version numbers (without relying on REGEX to support older CMake versions <3.9)
+string(SUBSTRING ${GNU_ASSEMBLER_VERSION} 0 ${DOT_POSITION} GNU_ASSEMBLER_MAJOR_VERSION)
+string(SUBSTRING ${GNU_ASSEMBLER_VERSION} ${DOT_POSITION} -1 DOT_AND_MINOR_VERSION)
+string(SUBSTRING ${DOT_AND_MINOR_VERSION} 1 -1 GNU_ASSEMBLER_MINOR_VERSION)
+math(EXPR GNU_ASSEMBLER_VERSION_NUMBER  "${GNU_ASSEMBLER_MAJOR_VERSION} * 1000 + ${GNU_ASSEMBLER_MINOR_VERSION}")
+set(TBB_COMMON_COMPILE_FLAGS ${TBB_COMMON_COMPILE_FLAGS} "-DGNU_AS_VERSION=${GNU_ASSEMBLER_VERSION_NUMBER}")
+message(STATUS "GNU Assembler version: ${GNU_ASSEMBLER_VERSION}  (${GNU_ASSEMBLER_VERSION_NUMBER})")
+
 # Enable Intel(R) Transactional Synchronization Extensions (-mrtm) and WAITPKG instructions support (-mwaitpkg) on relevant processors
 if (CMAKE_SYSTEM_PROCESSOR MATCHES "(AMD64|amd64|i.86|x86)")
     set(TBB_COMMON_COMPILE_FLAGS ${TBB_COMMON_COMPILE_FLAGS} -mrtm $<$<AND:$<NOT:$<CXX_COMPILER_ID:Intel>>,$<NOT:$<VERSION_LESS:${CMAKE_CXX_COMPILER_VERSION},11.0>>>:-mwaitpkg>)

--- a/cmake/compilers/GNU.cmake
+++ b/cmake/compilers/GNU.cmake
@@ -43,7 +43,7 @@ endif()
 # convert it to an integer that can be used in the C++ code to compare
 # against, and disable the __TBB_WAITPKG_INTRINSICS_PRESENT macro.
 # Unfortunately, binutils only reports the MAJOR.MINOR version, without patch,
-# so the check we effectively implement is version >2.32 (instead of >=2.31.1).
+# so the check we effectively implement is version >=2.32 (instead of >=2.31.1).
 # Capturing the output in CMake can be done like below. The version information
 # will written to either stdout or stderr; for me it's stderr, but let's just
 # grab both to be sure.

--- a/cmake/compilers/GNU.cmake
+++ b/cmake/compilers/GNU.cmake
@@ -36,16 +36,17 @@ if (NOT CMAKE_GENERATOR MATCHES "Ninja" AND NOT CMAKE_CXX_DEPENDS_USE_COMPILER)
 endif()
 
 
-# Binutils < 2.31.1 do not support the tpause instruction.
-# When compiling with a modern version of GCC (supporting it) but
-# relying on an outdated assembler, the assembler fails.
-# The following code invokes the GNU assembler to get the version number and
-# convert it to an integer that can be used in the C++ code to compare
-# against, and disable the __TBB_WAITPKG_INTRINSICS_PRESENT macro.
-# Binutils only report the version in the MAJOR.MINOR format, 
-# therefore the version checked is >=2.32 (instead of >=2.31.1).
-# Capturing the output in CMake can be done like below. The version information
-# is written to either stdout or stderr. Here, both are used. 
+# Binutils < 2.31.1 do not support the tpause instruction. When compiling with
+# a modern version of GCC (supporting it) but relying on an outdated assembler,
+# will result in an error reporting "no such instruction: tpause".
+# The following code invokes the GNU assembler to extract the version number
+# and convert it to an integer that can be used in the C++ code to compare
+# against, and conditionally disable the __TBB_WAITPKG_INTRINSICS_PRESENT
+# macro if the version is incompatible. Binutils only report the version in the
+# MAJOR.MINOR format, therefore the version checked is >=2.32 (instead of
+# >=2.31.1). Capturing the output in CMake can be done like below. The version
+# information is written to either stdout or stderr. To not make any
+# assumptions, both are captured.
 execute_process(
     COMMAND ${CMAKE_CXX_COMPILER} -xc -c /dev/null -Wa,-v -o/dev/null
     OUTPUT_VARIABLE ASSEMBLER_VERSION_LINE_OUT

--- a/cmake/compilers/GNU.cmake
+++ b/cmake/compilers/GNU.cmake
@@ -60,7 +60,7 @@ string(REGEX REPLACE "GNU assembler version ([0-9]+)\\.([0-9]+).*" "\\2" _tbb_gn
 unset(ASSEMBLER_VERSION_LINE_OUT)
 unset(ASSEMBLER_VERSION_LINE_ERR)
 unset(ASSEMBLER_VERSION_LINE)
-message(TRACE "Extracted GNU asssembler version: major=${_tbb_gnu_asm_major_version} minor=${_tbb_gnu_asm_minor_version}")
+message(TRACE "Extracted GNU assembler version: major=${_tbb_gnu_asm_major_version} minor=${_tbb_gnu_asm_minor_version}")
 
 math(EXPR _tbb_gnu_asm_version_number  "${_tbb_gnu_asm_major_version} * 1000 + ${_tbb_gnu_asm_minor_version}")
 set(TBB_COMMON_COMPILE_FLAGS ${TBB_COMMON_COMPILE_FLAGS} "-D__TBB_GNU_ASM_VERSION=${_tbb_gnu_asm_version_number}")

--- a/include/oneapi/tbb/detail/_config.h
+++ b/include/oneapi/tbb/detail/_config.h
@@ -335,7 +335,7 @@
 
 #define __TBB_TSX_INTRINSICS_PRESENT (__RTM__ || __INTEL_COMPILER || (_MSC_VER>=1700 && (__TBB_x86_64 || __TBB_x86_32)))
 
-#define __TBB_WAITPKG_INTRINSICS_PRESENT ((__INTEL_COMPILER >= 1900 || (__TBB_GCC_VERSION >= 110000 && GNU_AS_VERSION > 2031) || __TBB_CLANG_VERSION >= 120000) \
+#define __TBB_WAITPKG_INTRINSICS_PRESENT ((__INTEL_COMPILER >= 1900 || (__TBB_GCC_VERSION >= 110000 && GNU_AS_VERSION >= 2032) || __TBB_CLANG_VERSION >= 120000) \
                                          && (_WIN32 || _WIN64 || __unix__ || __APPLE__) && (__TBB_x86_32 || __TBB_x86_64) && !__ANDROID__)
 
 /** Internal TBB features & modes **/

--- a/include/oneapi/tbb/detail/_config.h
+++ b/include/oneapi/tbb/detail/_config.h
@@ -335,7 +335,7 @@
 
 #define __TBB_TSX_INTRINSICS_PRESENT (__RTM__ || __INTEL_COMPILER || (_MSC_VER>=1700 && (__TBB_x86_64 || __TBB_x86_32)))
 
-#define __TBB_WAITPKG_INTRINSICS_PRESENT ((__INTEL_COMPILER >= 1900 || (__TBB_GCC_VERSION >= 110000 && GNU_AS_VERSION >= 2032) || __TBB_CLANG_VERSION >= 120000) \
+#define __TBB_WAITPKG_INTRINSICS_PRESENT ((__INTEL_COMPILER >= 1900 || (__TBB_GCC_VERSION >= 110000 && __TBB_GNU_ASM_VERSION >= 2032) || __TBB_CLANG_VERSION >= 120000) \
                                          && (_WIN32 || _WIN64 || __unix__ || __APPLE__) && (__TBB_x86_32 || __TBB_x86_64) && !__ANDROID__)
 
 /** Internal TBB features & modes **/

--- a/include/oneapi/tbb/detail/_config.h
+++ b/include/oneapi/tbb/detail/_config.h
@@ -1,5 +1,5 @@
 /*
-    Copyright (c) 2005-2023 Intel Corporation
+    Copyright (c) 2005-2024 Intel Corporation
 
     Licensed under the Apache License, Version 2.0 (the "License");
     you may not use this file except in compliance with the License.
@@ -335,7 +335,7 @@
 
 #define __TBB_TSX_INTRINSICS_PRESENT (__RTM__ || __INTEL_COMPILER || (_MSC_VER>=1700 && (__TBB_x86_64 || __TBB_x86_32)))
 
-#define __TBB_WAITPKG_INTRINSICS_PRESENT ((__INTEL_COMPILER >= 1900 || __TBB_GCC_VERSION >= 110000 || __TBB_CLANG_VERSION >= 120000) \
+#define __TBB_WAITPKG_INTRINSICS_PRESENT ((__INTEL_COMPILER >= 1900 || (__TBB_GCC_VERSION >= 110000 && GNU_AS_VERSION > 2031) || __TBB_CLANG_VERSION >= 120000) \
                                          && (_WIN32 || _WIN64 || __unix__ || __APPLE__) && (__TBB_x86_32 || __TBB_x86_64) && !__ANDROID__)
 
 /** Internal TBB features & modes **/

